### PR TITLE
mds: prevent too many request of mdslog for osd at once when the mds restarts (default 0, means no limited)

### PR DESCRIPTION
**Pull request by simon-rock**
Fixes: http://tracker.ceph.com/issues/40028
Signed-off-by: simon gao <simon29rock@gmail.com>

### CODE REVIEW

File: src/mds/MDLog.cc [hunk 1]
```diff
@@ -626,7 +626,12 @@ void MDLog::trim(int m)
     op_prio = CEPH_MSG_PRIO_HIGH;
 
   unsigned new_expiring_segments = 0;
-
+  
+  unsigned expiring_segments_size = g_conf()->mds_log_max_expiring_segments;
+  if(segments.size() > pre_segments_size && expiring_segments_size > 0)
+    expiring_segments_size = std::max(expiring_segments_size,static_cast<unsigned>(segments.size() - pre_segments_size));
+  pre_segments_size = segments.size();
```
Comment (ukernel) [inline src/mds/MDLog.cc]: pre_segments_size is not used

Comment (simon-rock) [reply src/mds/MDLog.cc]: It's used here.
expiring_segments_size = std::max(expiring_segments_size,static_cast<unsigned>

Comment (simon-rock) [reply src/mds/MDLog.cc]: It's used here.
expiring_segments_size = std::max(expiring_segments_size,static_cast

Comment (ukernel) [reply src/mds/MDLog.cc]: only set pre_segments_size when replay is done. decrease it (until it reach 0) when segments expired.

Comment (simon-rock) [reply src/mds/MDLog.cc]: ok

File: src/common/legacy_config_opts.h [hunk 2]
```diff
@@ -412,6 +412,7 @@ OPTION(mds_log_max_events, OPT_INT)
 OPTION(mds_log_events_per_segment, OPT_INT)
 OPTION(mds_log_segment_size, OPT_INT)  // segment size for mds log, default to default file_layout_t
 OPTION(mds_log_max_segments, OPT_U32)
+OPTION(mds_log_max_expiring_segments, OPT_U32)
```
Comment (batrick) [inline src/common/legacy_config_opts.h]: We don't add options to this file anymore. Please remove this change.

File: src/mds/MDLog.cc [hunk 3]
```diff
@@ -626,7 +626,14 @@ void MDLog::trim(int m)
     op_prio = CEPH_MSG_PRIO_HIGH;
 
   unsigned new_expiring_segments = 0;
-
+  
+  unsigned expiring_segments_size = g_conf()->mds_log_max_expiring_segments;
```
Comment (batrick) [inline src/mds/MDLog.cc]: Use: `g_conf().get_val<uint64_t>("mds_log_max_expiring_segments");`

File: src/common/legacy_config_opts.h (hunk 2 above)
Comment (simon-rock) [reply src/common/legacy_config_opts.h]: oh... sorry

File: src/mds/MDLog.cc (hunk 3 above)
Comment (simon-rock) [reply src/mds/MDLog.cc]: ok

File: src/mds/MDLog.cc [hunk 4]
```diff
@@ -626,7 +626,14 @@ void MDLog::trim(int m)
...
 
   unsigned new_expiring_segments = 0;
-
+  
+  unsigned expiring_segments_size = g_conf().get_val<uint64_t>("mds_log_max_expiring_segments");
+  if(segments.size() > pre_segments_size && expiring_segments_size > 0)
+    expiring_segments_size = std::max(expiring_segments_size,static_cast<unsigned>(segments.size() - pre_segments_size));
+  // reset pre_segments_size
+  if (segments.size() < max_segments)
+    pre_segments_size = 0;
```
Comment (ukernel) [inline src/mds/MDLog.cc]: no longer needed

File: src/mds/MDLog.cc [hunk 5]
```diff
@@ -626,7 +626,14 @@ void MDLog::trim(int m)
     op_prio = CEPH_MSG_PRIO_HIGH;
 
   unsigned new_expiring_segments = 0;
-
+  
+  unsigned expiring_segments_size = g_conf().get_val<uint64_t>("mds_log_max_expiring_segments");
+  if(segments.size() > pre_segments_size && expiring_segments_size > 0)
+    expiring_segments_size = std::max(expiring_segments_size,static_cast<unsigned>(segments.size() - pre_segments_size));
```
Comment (ukernel) [inline src/mds/MDLog.cc]: this code limits expiring segments when segments.size() <= pre_segments_size. This is not expected

File: src/mds/MDLog.cc [hunk 6]
```diff
@@ -627,6 +627,13 @@ void MDLog::trim(int m)
 
   unsigned new_expiring_segments = 0;
 
+  unsigned max_expiring_segments = 0;
+  if (pre_segments_size > 0){
+    max_expiring_segments = max_segments/2;
+    assert(segments.size() >= pre_segments_size);
+    max_expiring_segments = std::max(max_expiring_segments,static_cast<unsigned>(segments.size() - pre_segments_size));
```
Comment (ukernel) [inline src/mds/MDLog.cc]: std::max<unsigned>

otherwise, looks good

Comment (simon-rock) [reply src/mds/MDLog.cc]: update it. thanks.

**Comment by ukernel** [comment]
LGTM

**Comment by simon-rock** [comment]
@batrick what can I do to fix this issue, make check(arm64).

File: src/mds/MDLog.cc [hunk 7]
```diff
@@ -1460,9 +1473,10 @@ void MDLog::_replay_thread()
     if (mds->is_daemon_stopping()) {
       return;
     }
+    pre_segments_size = segments.size();  // get num of logs when replay is finished
     finish_contexts(g_ceph_context, waitfor_replay, r);  
   }
-
+
```
Comment (batrick) [inline src/mds/MDLog.cc]: Please remove this whitespace change. Use `git reset -p HEAD^`.

File: src/mds/MDLog.h [hunk 8]
```diff
@@ -124,6 +124,7 @@ class MDLog {
   map<uint64_t,LogSegment*> segments;
   set<LogSegment*> expiring_segments;
   set<LogSegment*> expired_segments;
+  unsigned pre_segments_size;            // the num of segments when the mds finished replay-journal, to calc the num of segments growing
```
Comment (batrick) [inline src/mds/MDLog.h]: `std::size_t pre_segments_size = 0;`

File: src/mds/MDLog.h [hunk 9]
```diff
@@ -207,6 +208,7 @@ class MDLog {
 		      mdsmap_up_features(0),
                       submit_mutex("MDLog::submit_mutex"),
                       submit_thread(this),
+                      pre_segments_size(0),
```
Comment (batrick) [inline src/mds/MDLog.h]: ... and remove this.

**Comment by simon-rock** [comment]
@batrick . I find that these errors are about bluefs, osd and messenger...

**Comment by batrick** [comment]
> @batrick . I find that these errors are about bluefs, osd and messenger...

You can ignore the make check (arm64) errors.

**Comment by batrick** [comment]
> LGTM

@ukernel please approve so the merge script gives credit.

**Comment by ukernel** [comment]
can't find approve button

**Pull request merged**